### ISSUE TITLE
Consider supporting in-memory locking above the distributed layer in multiplexing

### ISSUE THREAD

**Issue by madelson** [madelson/DistributedLock#38]
Right now, if the same process tries to claim a lock, we push this out to the distributed locking layer (e. g. SQL). 

We could reduce resource usage by first checking an internal lock (e. g. a `SemaphoreSlim`). We have to be very careful about how this interops with modes, though.

**Comment by dazinator** [comment]
Seems sensible. To go one step further, before pushing to the distributed lock layer, it might also be valuable to add a SystemDistributedLock layer in front of that again?

Is is a desire of this library to provide any kind of common IDistributedLock abstraction? Or will it just provide different concrete lock implementations and leave it for application developers to  build there own abstractions of needed? If it's the latter then I could see myself wanting to create my own CompositeDistributedLock abstraction, which wraps an inner list. When acquiring the lock it will try each in order from the list. I'd then have  Inmemory first in that list, then SystemDistributed then Sql. Returning an IDisposable. I know that now all concrete lock types will support the same features (upgradeable etc etc) so this CompositeDistributedLock abstraction that I create will need to have an inner list of ICompositeLock which have only the most basic Aquire() method. To that end it would help if each of the concrete lock implementations that you provide already implement some basic interface for only the common features.

**Comment by madelson** [comment]
> Question: is it a desire of this package to provide any kind of common IDistributedLock abstraction that all concrete providers will implement at a minimum?

This is something I'm honestly a bit unsure about and definitely eager to hear feedback. In my use-cases, distributed locks aren't easily substituted because the semantics are just different (for System vs. SQL, for example). On the other hand, I could see a world where someone wanted to plug in a different implementation for deploying to a cloud environment vs. local.

**Comment by dazinator** [comment]
Yeah I think its valuable, when debugging locally I like dependencies minimal so using in memory locks is perfect. When deployed to staging and production we'd want a proper distributed lock using sql or azure blob leases etc. The only way to do that without horrible branching all over the place in the code is to create a common abstraction over the two, and make the concrete implementation actually used configurable or based on DI etc.

**Comment by dazinator** [comment]
I think one basic scenario that many people might look for is basically a replacement to this code, that makes if truly global:

```csharp

lock(_lock)
{
// blah
}
```

The next might be variations of `Monitor.Enter` I.e so you can do the same as above but supply timeouts etc. I'd hope there would be a way to achieve that basica level of lock acquisition accross all distributed lock implementations but I havent looked closely.

**Comment by madelson** [comment]
Some thoughts on composite locks:

* This should be more complex than just take lock a, take lock b. For example, let's say we have N local waiters for a the lock and then the first one of those acquires the distributed lock. We shouldn't release that until all those N local waiters have gotten to hold the lock or given up (new local waiters that come in after we got the distributed lock shouldn't get to join in; otherwise we might hog the distributed lock indefinitely). The benefit of this system is that it prevents a service using composite locking from always losing out to a service that doesn't, and also decreases the number of distributed lock operations (more efficient). The downside is that it reduces fine-grained interleaving of lock requests between services. Note that for R/W locks where we have both Write and UpgradeableRead, we have to be careful that the underlying lock is the right type.

* We need to be careful with R/W locks if they have writer-jumps-reader behavior. We wouldn't want a scenario where writers are queued up behind a local read lock hold which is then queued up waiting for a distributed write lock hold to be released. We might want to only use the local lock for writes and upgradeable reads in order to prevent this.

**Comment by darkflame0** [comment]
> 我认为许多人可能会寻找的一个基本场景基本上是该代码的替换/替代，这使其成为全球性的：
> 
> ```cs
> lock(_lock)
> {
> // blah
> }
> ```
> 
> 下一个可能是`Monitor.Enter`Ie 的变体，因此您可以执行与上述相同的操作，但提供超时等。我希望有一种方法可以在所有分布式锁实现中实现锁获取的基本级别，但我没有仔细研究。如果有，那么这可能是它们之间通用接口的基础。


any progress? 
To be honest,I need a InMemoryDistributedLock

**Comment by madelson** [comment]
Hi @darkflame0 I am not actively working on this feature. Can you help me understand your use-case? There are a few different ones I've considered:
* In-memory named lock implementation where the named locks simplify coordination between different parts of the same application
* In-memory named lock implementation to stand in for a truly-distributed version while testing
* In-memory implementation *layered on top of* truly-distributed version to increase scalability (e.g. if 100 threads across 10 apps are waiting for the same named `SqlDistributedLock`, that's 100 database connections. However, if each thread first waited for its local in memory lock and only could claim the remote lock after acquiring that, then we'd only need 10 database connections).

What are you trying to do?

**Comment by darkflame0** [comment]
> 
> * In-memory named lock implementation where the named locks simplify coordination between different parts of the same application
> * In-memory named lock implementation to stand in for a truly-distributed version while testing

Mainly because of these two points,I need a infrastructure,it's in memory at the beginning,but can be distributed in the future.

**Comment by madelson** [comment]
@darkflame0 have you considered using `FileDistributedLock` or (if you are on windows) `EventWaitHandleDistributedLock`? These are both light-weight single-machine options.

**Comment by darkflame0** [comment]
@madelson  Maybe FileDistributedLock is a proper alternative,I will try it.
thanks.

**Comment by darkflame0** [comment]
@madelson I find out FileDistributedLock not support ReaderWriterLock and Semaphore ...

**Comment by madelson** [comment]
@darkflame0 yeah different technologies offer different capabilities; I only implement when I think I can offer something robust and performant on top of the particular technology.

Looking into it, I think we can build a reasonable reader-writer implementation with wait handles based on [this technique](https://stackoverflow.com/a/45745901/1142970). Would that be useful to you?

**Comment by darkflame0** [comment]
It's useful me.I'm on Windows.

**Comment by darkflame0** [comment]
but wait handles is unavailable if on linux.
I still think should have a InMemory Implementation.
It is beneficial to both test and single application

**Comment by madelson** [comment]
@darkflame0 I'm working on some prototypes for these. If I put out a prerelease version would you be interested in trying it out?

**Comment by madelson** [comment]
@darkflame0 ok prerelease is out (https://www.nuget.org/packages/DistributedLock.ProcessScoped/1.0.0-alpha01). Let me know if you get a chance to give it a try. This doesn't support composite locking yet, just the process-scoped named lock types you were asking about.
